Reduce build time by making some tests faster

tinwelint opened the pull request:
Mostly in kernel and "neo4j" components for now. More to come as we go.

tinwelint committed:
Converted most of HaBeanIT to unit test

tinwelint committed:
Faster BatchInsertTest and fixes a bug in BatchInserter

speedup comes from using a global batch inserter for most tests,
i.e. an inserter which lives for the whole test class.

Bug was that changing node labels so that there were dynamic label
records deleted would both write those to the store AND queue them
up for writing on the next flush. Both of those writes would add
the ids to the store free list and so later allocation would be
messed up, some records overwriting existing records.

tinwelint committed:
Faster LockingCompatibilityTestSuite

due to improved lock acquisition wait detection and lifecycle handling
of lock clients

tinwelint committed:
Faster RelationshipGroupDegragmenterTest

by running on smaller dataset and test verifies some assumptions about itself
which it previous assumed to be true.

tinwelint committed:
Faster TestRelationshipCount

by using one db per threshold config

tinwelint committed:
Faster RelationshipGroupStoreIT

by only creating necessary amount of reltypes.
also added a unit test for the same issue.

tinwelint committed:
Renames ConsistentPropertyReadsTest -> ConsistentPropertyReadsIT

tinwelint committed:
Faster MetaDataStoreTest

by using Race and not overkilling atomic testing. With the new code
the races could be reproduced with something like 98% accuracy
and yet runs in 1/30th of the time.

tinwelint committed:
Faster CheckPointerIntegrationTest

by not waiting unnecessarily for the job scheduling in neo4j,
but instead being more proactive.

tinwelint committed:
Faster TransactionTerminationTest

by using Race and having additional end condition after successful
number of iterations as well as time.

tinwelint committed:
Faaster TestConcurrentRelationshipChainLoadingIssue

runs until end time OR number of intervals which has it run
a lower number of intervals and so faster on faster machines.
This test was originally for an issue in the object cache,
which now is gone, but test could still be here.

tinwelint committed:
Faster StoreNodeRelationshipCursorTest

by using single NeoStores for all test

tinwelint committed:
Faster StorePropertyCursorTest

by using static NeoStores. Also corrects some bad deletion patterns
which ended up frieing dynamic records multiple times.

tinwelint committed:
Faster RecordFormatTest

by using static PageCache and also reduces iterations to 100k from 1M,
we'll get full coverage anyway statistically.

tinwelint committed:
Faster SchemaStorageTest

by sharing db and co-creating schema

tinwelint committed:
Faster IndexingAcceptanceTest

by using static db and separate labels

tinwelint committed:
Faster KernelTransactionsTest

by using more efficient lock wait assertions

tinwelint committed:
Faster NeoStoresIT

Also fixes it to actually test what it was written to test.

previously there was a race that was supposed to be tested,
but there were a couple of things wrong with the test itself:
- it didn't run reader and writer in parallel, rather sequentially
- it set the threads up run a lot of iteration, but only waited for
  them to complete for a maximum of 2sec which would make most
  of the test run after the control of those threads had been released
  (the executor left them running when shutting down).

Now Race and a decent number of iterations is used and the threads
run in parallel.

tinwelint committed:
Faster GraphDatabaseServiceTest

due to more precise monitor waiting.

tinwelint committed:
Faster LabelScanStoreIT

by using class rule db

tinwelint committed:
Faster IndexStatisticsTest

by running on smaller dataset although with ability to run with higher
(controlled with system property)

tinwelint committed:
Faster IndexTxStateLookupTest

by using class rule db

tinwelint committed:
Faster TestStartTransactionDuringLogRotation

due to external forcing of log instead of waiting for background writer
to write enough to run into threshold so that rotation happens.
Cuts test time down to 1/10th or so.

tinwelint committed:
Simplifies ReadIndexWritesUnderConcurrentLoadStressIT

and increases number of threads, which both saturates the system more and
makes the test faster because it caters to the commit scaling of neo4j

tinwelint committed:
Cleans up threads after RWLockTest

tinwelint commented:
@benbc wdyt?

benbc commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -487,6 +487,7 @@ public void start() throws IOException
             dependencies.satisfyDependency( storageEngine.storeReadLayer() );
             dependencies.satisfyDependency( logEntryReader );
             dependencies.satisfyDependency( storageEngine );
+            dependencies.satisfyDependency( transactionLogModule.checkPointing() );
```
This looks like a surprising change for test speed-up work.

tinwelint replied:
This is about exposing a way for tests to force a checkpoint, no biggie

MishaDemianenko replied:
i think you need this at all, for example because  https://github.com/neo4j/neo4j/blob/3.1/community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java#L464
hack, and there is for example CheckpointerSupplier that rely on checkpointer to be registered

tinwelint replied:
Oh cool, then it can probably be removed

benbc commented on community/kernel/src/main/java/org/neo4j/unsafe/batchinsert/internal/BatchInserterImpl.java:
```diff
@@ -750,8 +750,7 @@ protected PropertyBlock underlyingObjectToObject( Entry<String, Object> property
         NodeLabels nodeLabels = parseLabelsField( nodeRecord );
-        nodeStore.updateDynamicLabelRecords( nodeLabels.put( getOrCreateLabelIds( labels ), nodeStore,
-                nodeStore.getDynamicLabelStore() ) );
+        nodeLabels.put( getOrCreateLabelIds( labels ), nodeStore, nodeStore.getDynamicLabelStore() );
```
Is this a semantic change? You should probably get someone from the Kernel team to review this as well as me.

tinwelint replied:
If you look at this change in the context of the specific commit, there's a description about the problem and that it fixes this bug

benbc commented on community/kernel/src/test/java/org/neo4j/graphdb/GraphDatabaseServiceTest.java:
```diff
@@ -71,44 +75,45 @@ public void givenDatabaseAndStartedTxWhenShutdownThenWaitForTxToFinish() throws
+            db.shutdown();
+            return null;
+        } );
+        t3.get().waitUntilWaiting( location -> location.isAt( DatabaseAvailability.class, "shutdown" ) );
```
Does this constitute an assertion that the shutdown is actually blocked by the in-flight transaction?

tinwelint replied:
Yes exactly, this technique has existed for years in the neo4j test code base, although many tests don't take advantage of it, instead waiting an arbitrary amount of seconds, catch a TimeoutException and interpret that is the same result. This technique is actually better in every way because we can very specifically determine where the wait happens and it can be detected much faster.

benbc commented on community/kernel/src/test/java/org/neo4j/graphdb/GraphDatabaseServiceTest.java:
```diff
@@ -71,44 +75,45 @@ public void givenDatabaseAndStartedTxWhenShutdownThenWaitForTxToFinish() throws
+        {
+            // expected
+        }
+        shutdownFuture.get();
```
Shouldn't this (and other `get()`s) have a timeout?

tinwelint replied:
We could, but there's no legit reason for shutdown to be blocked here, just as there's no legit reason for a normal shutdown of a db which happens in most of our tests to be blocked.

benbc commented on community/kernel/src/test/java/org/neo4j/graphdb/GraphDatabaseServiceTest.java:
```diff
@@ -172,81 +173,44 @@ public void terminateNestedTransactionThrowsExceptionOnNextNestedOperation() thr
     }
 
     @Test
     public void givenDatabaseAndStartedTxWhenShutdownAndStartNewTxThenBeginTxTimesOut() throws Exception
```
Shouldn't this be `BeginTxThrowsAnException`?

benbc commented on community/kernel/src/test/java/org/neo4j/graphdb/GraphDatabaseServiceTest.java:
```diff
@@ -172,81 +173,44 @@ public void terminateNestedTransactionThrowsExceptionOnNextNestedOperation() thr
         }
-
-        assertThat( result.get(), instanceOf( DatabaseShutdownException.class ) );
     }
```
My understanding from the test name is that this is intended to test the case where a transaction starts while the shutdown is in progress. So I would expect the `beginTx()` to come before the `barrier.release()`. Otherwise I don't see any reason to have the blocking tx in this test -- you  might as well just do a simple `db.shutdown()` followed by `db.beginTx()`.

tinwelint replied:
I thought so too, but the test looks made for a very specific bug perhaps and so didn't want to change something I didn't understand.

benbc commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/KernelTransactionsTest.java:
```diff
@@ -472,41 +475,28 @@ private static TransactionCommitProcess newRememberingCommitProcess( final Trans
+//                return kernelTransactions.newInstance( KernelTransaction.Type.explicit, AccessMode.Static.WRITE, 0L );
+//            }
+//        } );
+//    }
```
Can this just be deleted?

tinwelint replied:
Deleted now

benbc commented on community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointerIntegrationTest.java:
```diff
@@ -106,25 +112,46 @@ public void shouldCheckPointBasedOnTime() throws Throwable
+        while ( !checkPointInTxLog( db ) )
+        {
+            Thread.sleep( millis );
+        }
```
Does this need a timeout in case checkpointing is completely broken somehow?

tinwelint replied:
Yup, added

benbc commented on community/kernel/src/test/java/org/neo4j/test/OtherThreadExecutor.java:
```diff
@@ -217,17 +220,45 @@ public String toString()
+    public WaitDetails waitUntilWaiting( Predicate<WaitDetails> correctWait ) throws TimeoutException
+    {
+        return waitUntilThreadState( details -> true, Thread.State.WAITING, Thread.State.TIMED_WAITING );
+    }
```
ignores the predicate

tinwelint replied:
Yes, fixed

benbc commented on community/kernel/src/test/java/org/neo4j/test/OtherThreadExecutor.java:
```diff
@@ -217,17 +220,45 @@ public String toString()
+
+    public WaitDetails waitUntilBlocked( Predicate<WaitDetails> correctWait ) throws TimeoutException
+    {
+        return waitUntilThreadState( details -> true, Thread.State.BLOCKED );
```
Ignores the predicate

tinwelint replied:
Yes, fixed

benbc reviewed (commented):
I'm afraid my eyes glazed over a bit towards the end there. Looks like some good improvements, though. How much speed-up do you think there is overall?

benbc commented:
Please get someone from your team to review as well. I don't feel qualified to have an opinion on some of that stuff.

MishaDemianenko commented on community/io/src/test/java/org/neo4j/test/rule/TestDirectory.java:
```diff
@@ -246,8 +246,31 @@ private void evaluateClassBaseTestFolder( )
+     * This method can be used outside the context of a Rule as a utility. It will return something like:
+     * {@code component/target/test-data/org.neo4j.location.of.my.test.Clazz/}
+     */
+    public static File testDataDirectoryOf( FileSystemAbstraction fs, Class<?> owningTest, boolean clean )
```
all methods that create test directory is called testDirectory(something...) this static method should be called in the same way

tinwelint replied:
It's not a rule factory method and so wanted to call it something else, actually it should be called something very different to not be intermixed with the common rule factory methods, shouldn't it?

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/unsafe/batchinsert/internal/BatchInsertTest.java:
```diff
@@ -1722,6 +1559,19 @@ private void setAndGet( BatchInserter inserter, Object value )
+
+    private void forceFlush( BatchInserter inserter )
+    {
+        ((BatchInserterImpl)inserter).getNeoStores();
```
this is extremely confusing. We expose method forceFlushChanges to flush changes

tinwelint replied:
Alright, the thinking was that you get NeoStores to look at records things and when you do you want all changes to be present. Sure this method could be used to get the NeoStores at after creation and then just use it. I see your point.

MishaDemianenko commented on community/io/src/test/java/org/neo4j/test/rule/TestDirectory.java:
```diff
@@ -246,8 +246,31 @@ private void evaluateClassBaseTestFolder( )
         File testData = new File( locateTarget( owningTest ), "test-data" );
-        testClassBaseFolder = new File( testData, owningTest.getName() ).getAbsoluteFile();
+        File result = new File( testData, owningTest.getName() ).getAbsoluteFile();
+        if ( clean )
```
there is already method cleanDirectory that with a bit of adaptation will be suitable for static and general case

tinwelint replied:
OK, slightly beyond the scope of this PR, but will extract this method anyway

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/batchinsert/internal/BatchInserterImpl.java:
```diff
@@ -1102,6 +1103,18 @@ private void dumpConfiguration( Map<String,String> config, PrintStream out )
+    // test-access
+    NeoStores getNeoStores()
+    {
+        forceFlushChanges();
```
this is confusing i think it's up to test method to do explicit call to flush changes instead.

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/unsafe/batchinsert/internal/BatchInsertTest.java:
```diff
@@ -188,13 +192,37 @@ public BatchInsertTest( int denseNodeThreshold )
+        // Global inserter can be used in tests which simply want to verify "local" behaviour,
+        // e.g. create a node with some properties and read them back.
+        globalInserter = BatchInserters.inserter(
+                TestDirectory.testDataDirectoryOf( REAL_FS, BatchInsertTest.class, true ),
```
fs should be used here right?

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/locking/StopCompatibility.java:
```diff
@@ -556,16 +554,9 @@ private void assertNoLocksHeld()
     private void assertThreadIsWaitingForLock( LockAcquisition lockAcquisition ) throws Exception
     {
-        for ( int i = 0; i < 30; i++ )
+        for ( int i = 0; i < 30 && !suite.isAwaitingLockAcquisition( lockAcquisition.executor.waitUntilWaiting() ); i++ )
```
that's not equivalent behaviour to what we have before since before we where checking that we actually waiting for a lock for some time now it can happen that we just in a method that was expected.

tinwelint replied:
It is equivalent and even asserts more than we did previously. In this piece of code we wait for the thread to go into a _waiting_ state AND see that it waits in the correct place. The previous code only verified that the thread hadn't acquired the lock in N seconds, it could have been caused by anything.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/util/DebugUtil.java:
```diff
@@ -364,4 +368,78 @@ private static boolean hasTestAnnotation( Method method )
+     * Super simple utility for determining where most time is spent when you don't know where to even start.
+     * It could be used to home in on right place in a test or in a sequence of operations or similar.
+     */
+    public abstract static class Timer
```
Those kind of things are called `Stopwatch` and we have already one from Guava and from apache on our test class path do not see why we need something else...

tinwelint replied:
How is Stopwatch the same thing? This simple utility prints times at different points where it's used. It's the printing which is useful and it's only a utility quite useful when seeing what takes time in a test.

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/MetaDataStoreTest.java:
```diff
@@ -399,22 +391,33 @@ public void testRecordTransactionClosed() throws Exception
+            AtomicLong fileReadCount = new AtomicLong();
+            AtomicLong apiReadCount = new AtomicLong();
+            int limit = 10_000;
+            long endTime = currentTimeMillis() + SECONDS.toMillis( 1 );
```
with that small endTime it's really big chance that on CI it will just will not continue based on endTime, i think we need to wait for at least of couple writes to happen

tinwelint replied:
Right we could perhaps increase that to 10 seconds or something, just to be safe AND put a lower bound of at least 100 each.

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/MetaDataStoreTest.java:
```diff
@@ -595,25 +562,16 @@ public void transactionClosedMustBeAtomic() throws Exception
-            }
+                apiReadCount.incrementAndGet();
+            } ) );
+            race.go();
```
same as above

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/MetaDataStoreTest.java:
```diff
@@ -542,46 +508,47 @@ public void transactionCommittedMustBeAtomic() throws Exception
-            {
-                future.get(); // We assert that this does not throw
-            }
+            race.go();
```
same as above here, we need to wait for couple of expected events to happen at least, too aggressive end time can just cancel race

MishaDemianenko reviewed (commented):
@tinwelint  Review round completed please find comments on appropriate lines.
Also general comment: I do not find `Race` thingy readable in most of the cases more confusing....
Also in most of the races place previous assertions just gone and we check only that race did not fail which is strange - we should check at least that our desired state/condition was achieved at least once, also having some assertions during the race should clarify situation.

tinwelint committed:
Extracted cleanup method in TestDirectory

tinwelint committed:
Ensures MetaDataStoreTest race tests runs at least some operations

even on really slow machines. Also corrected some confusing test access
methods.

tinwelint commented:
@MishaDemianenko please tell me more about your concerns with `Race`, imho it makes race testing small, consice and less error prone than any other means I've seen in our testing. Also where have I accidentally removed assertions?

tinwelint committed:
Removes accessor of TestDirectory outside Rule context

tinwelint committed:
Slighty revamped Race API and tests for it

MishaDemianenko commented:
@tinwelint looks good now, will merge on green
